### ISSUE TITLE
How to focus on primary button when modal loads?

### ISSUE THREAD

**Issue by kevhou** [rackt/react-modal#51]
I can't seem to use jquery .focus() inside componentdidmount.

Someone please help!

**Comment by olegshilov** [comment]
@kevhou try using [Refs](https://facebook.github.io/react/docs/more-about-refs.html#completing-the-example)

**Comment by arasmussen** [comment]
I'm having a similar problem with not being able to .focus something inside componentDidMount. I'm using refs.

```javascript
var VIEWER_REF = 'VIEWER';
var PhotoViewer = React.createClass({
  componentDidMount: function() {
    this.refs[VIEWER_REF].getDOMNode().focus();
  },
  render: function() {
    return (
      <div tabIndex="0" ref={VIEWER_REF}>
        {contents}
      </div>
    );
  },
});
```

I've verified that this.refs[VIEWER_REF].getDOMNode().focus() is defined and being called. It worked before I put my PhotoViewer inside react-modal.

**Comment by arasmussen** [comment]
Also worth noting that document.activeElement (currently focused element) after my PhotoViewer opens is the ReactModal__Content element, so it seems like it's stealing focus from me. Why does the modal need focus?

**Comment by arasmussen** [comment]
Can we get a prop <Modal allowFocus={false} /> (default true) so that we can disable this functionality? I don't see why the Modal needs focus. I'm happy to manually pipe through ESC to changing isOpen to false.

**Comment by arasmussen** [comment]
Also sorry for the quadruple post, but a temporary workaround is the following:

```javascript
componentDidMount: function() {
  setTimeout(function() {
    this.refs[MY_REF].getDOMNode().focus();
  }.bind(this), 0);
},
```

**Comment by brandly** [comment]
In case anyone else comes across this, as of `v0.13`, [`getDOMNode()` is deprecated.](https://facebook.github.io/react/blog/2015/03/10/react-v0.13.html#new-features) The above example would be changed to this:

```js
componentDidMount: function() {
  setTimeout(function() {
    React.findDOMNode(this.refs[MY_REF]).focus();
  }.bind(this), 0);
},
```

**Comment by SebT** [comment]
I'm having a similar problem, but I just can't get my ref.

```javascript

render() {
  return (
    <Modal>
        <input ref="input"/>
    </Modal>
}

componentDidMount() {
    console.log(this.refs.input); // undefined
}
```

I also tried to add  `ref="modal"` on the Modal element and access the input with `this.refs.modal.refs.input` but it's undefined, yet `this.refs.modal` isn't.

If I can't get the ref to my input, it's even harder to focus it x) 

Any idea? I'm using react 0.13.3 et react-modal 0.3.0

**Comment by brandly** [comment]
if you want to focus on an `input`, you should just be able to put `autoFocus` on it, unless you need more control over it

**Comment by tazsoulis** [comment]
i have the same problem http://prntscr.com/8f6cve

**Comment by SebT** [comment]
autoFocus={true} does not work and I still can't get a ref to my input (see 2 posts above), any help please?

**Comment by brandly** [comment]
i think you just put `<input autoFocus />`. [source](http://stackoverflow.com/a/28891210/1614967)

**Comment by tomhirschfeld** [comment]
I'm also having the issue where the react-modal becomes the active element upon opening and I need a way to prevent this from happening.

**Comment by SebT** [comment]
`autoFocus` doesn't work because it focuses the input when the modal is mounted. 

We need to focus the input when the modal opens:

```javascript
componentDidUpdate: function(prevProps, prevState) {
  if (prevState.yourVar !== this.state.yourVar && this.state.yourVar === true) {
    setTimeout(function() {
      React.findDOMNode(this.refs.yourInput).focus();
    }.bind(this), 0);
  }
}
```
`yourVar` is the variable you give to the modal as `isOpen`.

This works perfectly fine for me.

**Comment by wader** [comment]
@SebT Thanks. Do you know why `setTimeout` is needed? if i log in a ref callback i see that `componentDidUpdate` is called before the ref callback... is that how it is suppose to work?

**Comment by ryanflorence** [comment]
Let's say you are a screen reader user, when the modal opens, and you folks go and focus the primary button, how does the user know what the content of the modal says? They don't. They also don't know if they should move the virtual cursor up or down to find out.

I can't remember why it's in a setTimeout, but I guess it's nice that it keeps you from confusing blind folks :P

If there's a good accessibility argument here, I'll reopen, but as far as I can tell we are doing the most accessible thing, which is the primary goal of this component.

**Comment by vvlevykin** [comment]
@ryanflorence can you please point to a guide that says this?

This article contradicts what you said: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_dialog_role#Focus_management

**Comment by kpdecker** [comment]
@ryanflorence It seems to me that @vvlevykin has a compelling a11y argument for at least allowing this behavior to be conditionally controlled. For use cases where we do need to have this behavior, i.e. a properly labeled and annotated form, we have to either have a focus+blur+focus cycle that causes validation issues under redux-forms, or we have to play the game of using setTimeout to avoid a race condition with this code and hoping that it works in all situations.

**Comment by gertig** [comment]
Thanks @SebT for the hack around the setTimeout!

@ryanflorence can you explain why forcing the content to be focused is preferred over allowing autoFocus to work on inputs?

**Comment by bl-nero** [comment]
This bites me heavily on iOS/Safari. The Safari developers decided to be smarter than everyone, and decided that you can't focus() an input anytime else than when consuming a click event. This means that the setTimeout() hack won't work there. :(

**Comment by bl-nero** [comment]
@ryanflorence Would you consider reopening this bug because of what I wrote above?

**Comment by bl-nero** [comment]
@ryanflorence Ping. This is still a problem, and I still haven't found a compelling solution (other than not using react-modal in the first place).

**Comment by fritz-c** [comment]
`componentDidMount`/`componentDidUpdate` weren't working for me, but then I used the `onAfterOpen` prop as described in [this PR's](https://github.com/reactjs/react-modal/pull/154) code (using the ref callback rather than the deprecated string notation they used), and I was able to focus on a text input.

**Comment by dspnorman** [comment]
Please consider re-opening. This is still a major issue with this library. Per @vvlevykin a11y best practice is to focus the primary interactive element in the modal upon open. Stealing focus to the modal container actually breaks the experience for users of assistive technology.

**Comment by dspnorman** [comment]
This appears to be fixed in #222
